zkevm-circuits/src/evm_circuit/util/common_gadget.rs

```diff
@@ -445,3 +452,220 @@ impl<F: Field> TransferGadget<F> {
         Ok(())
     }
 }
+
+#[derive(Clone, Debug)]
+pub(crate) struct CommonCallGadget<F, const CALL_OP: bool> {
+    pub is_success: Cell<F>,
+
+    pub gas: Word<F>,
+    pub gas_is_u64: IsZeroGadget<F>,
+    pub callee_address: Word<F>,
+    pub value: Word<F>,
+    pub cd_address: MemoryAddressGadget<F>,
+    pub rd_address: MemoryAddressGadget<F>,
+    pub memory_expansion: MemoryExpansionGadget<F, 2, N_BYTES_MEMORY_WORD_SIZE>,
+
+    value_is_zero: IsZeroGadget<F>,
+    pub has_value: Expression<F>,
+    pub callee_code_hash: Cell<F>,
+    pub is_empty_code_hash: IsEqualGadget<F>,
+}
+
+impl<F: Field, const CALL_OP: bool> CommonCallGadget<F, CALL_OP> {
+    pub(crate) fn construct(
+        cb: &mut ConstraintBuilder<F>,
+        is_call: Expression<F>,
+        is_callcode: Expression<F>,
+        is_delegatecall: Expression<F>,
+    ) -> Self {
+        let gas_word = cb.query_word();
+        let callee_address_word = cb.query_word();
+        let value = cb.query_word();
+        let cd_offset = cb.query_cell();
+        let cd_length = cb.query_rlc();
+        let rd_offset = cb.query_cell();
+        let rd_length = cb.query_rlc();
+        let is_success = cb.query_bool();
+
+        // Lookup values from stack
+        // `callee_address` is poped from stack and used to check if it exists in
+        // access list and get code hash.
+        // For CALLCODE, both caller and callee addresses are `current_callee_address`.
+        // For DELEGATECALL, caller address is `current_caller_address` and
+        // callee address is `current_callee_address`.
+        // For both CALL and STATICCALL, caller address is
+        // `current_callee_address` and callee address is `callee_address`.
+        cb.stack_pop(gas_word.expr());
+        cb.stack_pop(callee_address_word.expr());
+
+        if CALL_OP {
+            // `CALL` opcode has an additional stack pop `value`.
+            cb.condition(is_call + is_callcode, |cb| cb.stack_pop(value.expr()));
+        } else {
+            cb.stack_pop(value.expr());
+        }
+        cb.stack_pop(cd_offset.expr());
+        cb.stack_pop(cd_length.expr());
+        cb.stack_pop(rd_offset.expr());
+        cb.stack_pop(rd_length.expr());
+        if CALL_OP {
```

Comment:
The `CALL_OP` arg is set to false in `error_oog_call.rs` and set to true in `callop.rs`.
We should rename `CALL_OP` to `IS_ERROR` because that differentiates `callop.rs` and  `error_oog_call.rs`.

Reply:
fixed in [c552d5f](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1023/commits/c552d5f6f74350278101b64bd0ae7f04f21226cb)